### PR TITLE
GEOMESA-1330 Replacing SimpleFeatureTypes spec parsing with parboiled…

### PR DESCRIPTION
**Pull request by elahrvivaz**
… implementation

* All simple feature type user data are strings for consistency
* Separated and consolidated configuration constants into different objects in SimpleFeatureTypes

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by jnh5y** [comment]
@Afterglow375 @aheyne mind reading through this one?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala [hunk 1]
```diff
@@ -307,43 +302,38 @@ object RichSimpleFeatureType {
...
-      sft.getUserData.update(KEYWORDS_KEY, remainingKeywordsString)
-    }
+    def kv = k ~ kvSep ~ v ^^ { case key ~ kvSep ~ value => key -> value }
+    def kvList = repsep(kv, pairSep) ^^ { case x => x.toMap }
 
-    def removeAllKeywords() : Unit = {
-      sft.getUserData.remove(KEYWORDS_KEY)
+    def parse(s: String): Map[String, String] = {
+      parse(kvList, s.trim) match {
+        case Success(t, r)     if r.atEnd => t
```
Comment (Afterglow375) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala]: t and r? unclear variable names.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpec.scala [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+/**
+  * Intermediate format for attribute descriptors
+  */
+sealed trait AttributeSpec {
+
+  /**
+    * Attribute name
+    *
+    * @return
```
Comment (Afterglow375) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpec.scala]: Fill out the `@return` annotations?

**Comment by Afterglow375** [comment]
Looks fine to me.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpec.scala (hunk 2 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpec.scala]: fixed

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala (hunk 1 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala]: fixed

File: geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/AvroSimpleFeatureUtils.scala [hunk 3]
```diff
@@ -118,13 +116,10 @@ object AvroSimpleFeatureUtils {
         case t if classOf[Geometry].isAssignableFrom(t) =>
           (v: AnyRef) => ByteBuffer.wrap(wkbWriter.write(v.asInstanceOf[Geometry]))
 
-        case t if ad.isList => (v: AnyRef) =>
-          encodeList(v.asInstanceOf[java.util.List[_]],
-            ad.getUserData.get(USER_DATA_LIST_TYPE).asInstanceOf[Class[_]])
+        case t if ad.isList => (v: AnyRef) => encodeList(v.asInstanceOf[java.util.List[_]], ad.getListType())
```
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-avro/src/main/scala/org/locationtech/geomesa/features/avro/AvroSimpleFeatureUtils.scala]: fixed

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpecParser.scala [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  // any attribute
+  def attribute: Rule1[AttributeSpec] = rule {
+    simpleAttribute | geometryAttribute | listAttribute | mapAttribute
+  }
+
+  // builds a SimpleAttributeSpec for primitive types
+  private def simpleAttribute: Rule1[AttributeSpec] = rule {
+    (name ~ ":" ~ simpleType ~ attributeOptions) ~~> SimpleAttributeSpec
```
Comment (jnh5y) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureSpecParser.scala]: Insane request:  
Is it possible to handle attribute names with ":"'s in them?

**Pull request merged**